hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/CarbonitedMetadataThawingActor.scala

```diff
@@ -2,96 +2,148 @@ package cromwell.services.metadata.hybridcarbonite
       stop()
   }
 
   whenUnhandled {
+    case Event(Status.Failure(failure), data: WorkingData) =>
+      data.requester ! FailedMetadataResponse(data.action, failure)
+      stop()
     case other =>
       log.error(s"Programmer Error: Unexpected message to ${self.path.name}: $other")
       stay()
   }
-
-  def mergeResponses(metadata: Json, labels: Map[WorkflowId, Map[String, String]]): String = JsonEditor.updateLabels(metadata, labels).printWith(Printer.spaces2)
-
 }
 
 object CarbonitedMetadataThawingActor {
 
   def props(carboniterConfig: HybridCarboniteConfig, serviceRegistry: ActorRef, ioActor: ActorRef): Props =
     Props(new CarbonitedMetadataThawingActor(carboniterConfig, serviceRegistry, ioActor))
 
-  sealed trait ThawCarboniteMessage
-
-  final case class ThawCarbonitedMetadata(workflowId: RootWorkflowId) extends ThawCarboniteMessage
-  final case class ThawCarboniteSucceeded(value: String) extends ThawCarboniteMessage
-  final case class ThawCarboniteFailed(reason: Throwable) extends ThawCarboniteMessage
-
   final case class GcsMetadataResponse(response: Json)
 
-  sealed trait CarbonitedMetadataThawingState
-  case object CarbonitedMetadataThawingActorPendingState extends CarbonitedMetadataThawingState
-  case object CarbonitedMetadataThawingActorRunningState extends CarbonitedMetadataThawingState
+  sealed trait ThawingState
+  case object PendingState extends ThawingState
+  case object NeedsGcsAndLabelsState extends ThawingState
+  case object NeedsOnlyGcsState extends ThawingState
 
-  sealed trait CarbonitedMetadataThawingData
-  case object CarbonitedMetadataThawingActorPendingData extends CarbonitedMetadataThawingData
-  sealed trait CarbonitedMetadataThawingActorWorkingData extends CarbonitedMetadataThawingData { def requester: ActorRef }
-  final case class CarbonitedMetadataThawingActorDataNothingYet(requester: ActorRef) extends CarbonitedMetadataThawingActorWorkingData
-  final case class CarbonitedMetadataThawingActorDataWithGcsResponse(requester: ActorRef, gcsResponse: Json) extends CarbonitedMetadataThawingActorWorkingData
-  final case class CarbonitedMetadataThawingActorDataWithLabels(requester: ActorRef, labels: Map[WorkflowId, Map[String, String]]) extends CarbonitedMetadataThawingActorWorkingData
+  sealed trait ThawingData
+  case object PendingData extends ThawingData
+  sealed trait WorkingData extends ThawingData {
+    def requester: ActorRef
+    def action: WorkflowMetadataReadAction
+  }
+  final case class WorkingNoData(requester: ActorRef, action: WorkflowMetadataReadAction) extends WorkingData
+  final case class WorkingSomeData(requester: ActorRef, action: WorkflowMetadataReadAction, gcsResponse: Option[Json] = None, labels: Option[Map[WorkflowId, Map[String, String]]] = None) extends WorkingData
+
+  implicit class EnhancedJson(val json: Json) extends AnyVal {
+    def editFor(action: WorkflowMetadataReadAction): Json = action match {
```

Comment:
This `editFor` method looks like it could do with some comprehensive unit testing (in addition to any sub-unit testing going on in `JsonEditor`)

Reply:
This is very much WIP, perhaps once things stabilize and we have a firm understanding of what's required for the call-level stuff.